File: 0053-maximum-subarray/0053-maximum-subarray.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        if nums is None or len(nums) == 0:
            return 0

        result = nums[0]

        rSum = nums[0]

        for i in range(1,len(nums)):

            rSum =  max(rSum + nums[i], nums[i])

            result = max(result, rSum)

        return result

#time complexity: O(n)
#space complexity: O(1)

File: 0053-maximum-subarray/0053-maximum-subarray.java
class Solution {
    public int maxSubArray(int[] nums) {
        if(nums == null || nums.length == 0){
            return 0;
        }

        int result = nums[0];
        int n = nums.length;
        int prev  = Math.max(nums[0], 0);
        for(int i = 1; i < n; i++){

            prev = Math.max(nums[i], prev + nums[i]);
            result = Math.max(result, prev);
        }

        return result;
    }

}

// Time Complexity: O(n)
// Space Complexity: O(1)